### ISSUE TITLE
Discussion: interesting task-queue usage

### ISSUE THREAD

**Issue by jods4** [aurelia/task-queue#13]
So I was reviewing Aurelia code for #12 and this piece of code, found in `repeat.js` got me thinking:

``` javascript
handleInnerCollectionMutated(collection, changes) {
  // guard against source expressions that have observable side-effects that could
  // cause an infinite loop- eg a value converter that mutates the source array.
   if (this.ignoreMutation) {
    return;
  }
  this.ignoreMutation = true;
  let newItems = this.sourceExpression.evaluate(this.scope, this.lookupFunctions);
  this.observerLocator.taskQueue.queueMicroTask(() => this.ignoreMutation = false);
   // call itemsChanged...
  if (newItems === this.items) {
    // call itemsChanged directly.
    this.itemsChanged();
  } else {
    // call itemsChanged indirectly by assigning the new collection value to
    // the items property, which will trigger the self-subscriber to call itemsChanged.
    this.items = newItems;
  }
}
```

I understand the objective here but it doesn't feel right to me.

First it doesn't protect against _all_ circular changes. Because if a computed property is in the cycle this won't work, as the computed property itself will be queued _after_ the `ignoreMutation` reset.

Second, the problem at hand is not unique to this binding. In fact any code might trigger an evaluation that may have side effects that may create a cycle. Cycles in the dependencies are coding errors but they are sometimes unintended and it's good to have some protection against them. But I feel like the code above is too narrow and specific enough. To quote the comment: any binding can have a value converter.

Random thought: what about limiting the number of iterations in the micro-task loop?

**Comment by jdanyow** [comment]
I agree- a value converter that produces side-effects is a coding error.

The block of code you referenced is to protect against a common mistake where the value converter implementation looks like:

``` js
toView(array) {
  return array.sort(...);
}
```

I think limiting the number of iterations in the loop is a good idea, in the sense that it will make the system "fail fast" rather than hang the browser. It would not be a recovery mechanism due to the perf and bug-masking issues it would cause. Failing to fully flush the task-queue is a fatal error imho.

@EisenbergEffect thoughts?
@jods4 if you have some ideas on how the flush iteration cap might be implemented a PR is always appreciated

**Comment by EisenbergEffect** [comment]
@jdanyow I certainly want to help developers avoid errors and fail fast is a good characteristic. Do you have some ideas on what you would change?

**Comment by jdanyow** [comment]
I don't- was just saying to @jods4 that I liked the idea of having a cap because it would be a "fail fast" instead of "hang".  Either way it's a fatal error.  

@jods4 can this be closed or are you working on a change for this?

**Comment by jods4** [comment]
When I have some time I might tackle the task queue issues in #12. I think that would be the right place to implement an iteration limit.
Not sure if you want to keep this issue open/linked or close it.

If we add an iteration limit to the task queue it would be good to remove code such as this one, although it would be difficult to find its occurrences.

**Comment by jdanyow** [comment]
@jods4 any updates on this- should we keep it open?

**Comment by jods4** [comment]
@jdanyow as I wrote in #12, I got caught in work and was too busy to work on this :disappointed: 
I am traveling for 1 month so don't expect any PR from me any time soon.

This ticket is not critical so feel free to keep it open or close it as you see fit. As I wrote above it would probably be a good idea to implement such an iteration cap into the task queue itself, which could be done at the same time as the work suggested in #12.
